README.md

```diff
@@ -1,5 +1,213 @@
+@prefix p: <https://pipeline.described.at/> .
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
+```
+
+### Pipeline
+
+The pipeline itself must have a `rdf:type` of `p:Pipeline`.
+Based on the interfaces of the first and the last step, the pipeline itself is a [Writable](https://nodejs.org/api/stream.html#stream_writable_streams) and/or [Readable](https://nodejs.org/api/stream.html#stream_readable_streams) stream.
+Additional to the `p:Pipeline` `rdf:type`, `p:Writable` and/or `p:Readable` must be defined accordingly.
+That can look like this:
+
+```turtle
+<#pipeline> a p:Pipeline, p:Writable, p:Readable .
+```
+
+**Note:** The type of the streams could be also detected during runtime or based on metadata for the streams, but that would required deeper understanding of the pipeline and steps structures.
+To simplify handling of pipeline definitions without runtime data, the stream types are attached to the pipeline definition and errors are thrown in the runtime if they don't match. 
+
+### Step
+
+A pipeline consists of one or more steps, which are actually factories which return streams.
+Each step is linked with the previous one via [.pipe](https://nodejs.org/api/stream.html#stream_readable_pipe_destination_options).
+As this requires to process the steps in order, the steps must be provided as an RDF List.
+This example shows a very simple pipeline which consists of two steps: 
+
+```turtle
+<#pipeline> a p:Pipeline ;
+  p:steps [
+    p:stepList ( <#readFile> <#writeFile> )
+  ] .
+```
+
+**Note:** With the two levels (`p:steps` and `p:stepList`) it's possible to use a known identifier (Blank Node with name or Named Node) for a step list and reference to it in multiple pipelines.
+This is useful if pipelines share the same steps, but have different variables. 
+
+The step itself must have a `rdf:type` of `p:Step`.
+The Code Ontology is used to point to the actual code.
+This is done with the `code:implementedBy` property and code description with the `rdf:type` of `code:EcmaScript`.
+The `code:link` points to the JavaScript code file.
+`node:` IRIs can be used to use the Node.js resolve logic of `require` to find files and packages.
+`file:` IRIs will be expanded to full paths.
+This example uses the Node.js file system function to create a readable stream.
+`input.txt` is used as the only parameter.
+
+```turtle
+<#readFile> a p:Step ;
+  code:implementedBy [ a code:EcmaScript ;
+    code:link <node:fs#createReadStream>
+  ] ;
+  code:arguments ( "input.txt" ) .
+```
+
+It's also possible to give the arguments as key value pairs like this:
+
+```turtle
+<#parseCsvw> a p:Step ;
+  code:implementedBy [ a code:EcmaScript ;
+    code:link <file:customStreamCode.js>
+  ] ;
+  code:arguments [
+    p:name "url";
+    p:value "http://example.org/"
+  ], [
+    p:name "method";
+    p:value "POST"
+  ] .
+```
+
+The function in `customStreamCode.js` will be called with the following object:
```

Comment:
I would make it clear somewhere in this section that the implementation must export function, which takes loaded `code:arguments` and returns a `Stream`.